Timeout and Max-in-flight don't report non-resource URLs correctly.

smarterclayton opened the pull request:
Unify error reporting for 429 and 504 to be correct for timeout and max in flight and eviction. Add better messages to eviction (removing a todo). Return the correct body content for timeouts (reason and code should be correct).

This potentially increases cardinality of 429, but because non-api urls may be under the max-inflight budget we need to report them somewhere (if something breaks and starts fetching API versions endlessly).

```release-note
The 504 timeout error was returning a JSON error body that indicated it was a 500.  The body contents now correctly report a 500 error.
```

smarterclayton committed:
Update generic errors with the new http package codes

All of these errors are now part of the standard HTTP method. Formalize
those into our error types and remove duplication and unclear
separation.

smarterclayton committed:
Report non-resource URLs in max-in-flight correctly

This potentially has high cardinality, however we can rate limit based
on queries to these endpoints as well.

smarterclayton committed:
Timeout filter returns 504 and an inconsistent error body

Our rules are that code of the error must match code of the response. We
were also not setting the correct reason. This updates the timeout
filter to be consistent with other clients, without changing the error
code (504 is correct). The new message properly indicates the request
may still be running, which the old message did not do.

smarterclayton committed:
Return a status cause for disruption budget that contains more details

Also uses the standard error constructor for TooManyRequests and
clarifies *why* a disruption is rejected.

smarterclayton committed:
generated: bazel

deads2k commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:
```diff
@@ -211,6 +201,21 @@ func NewBadRequest(reason string) *StatusError {
+func NewTooManyRequests(message string, retryAfterSeconds int) *StatusError {
+	return &StatusError{metav1.Status{
+		Status:  metav1.StatusFailure,
+		Code:    http.StatusServiceUnavailable,
```
I don't keep them all in cache, but isn't this a 503, not a 429?

smarterclayton replied:
Um, yes, I don't know how this didn't get copied.  I must have autocompleted the wrong field.  Terrifying.

deads2k commented:
double check the status code I pointed out.  otherwise lgtm.

smarterclayton committed:
SuggestClientDelay is not about retrying, clarify message and header

SuggestClientDelay is returning whether the server has requested that
the client delay their next action. It is *not* about whether the client
should retry the action. Webhook was using it incorrectly, and the
method is now up to date.

smarterclayton committed:
generated: clarification on RetryAfterSeconds field

smarterclayton commented:
@deads2k see the SuggestClientDelay bit

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go:
```diff
@@ -90,6 +90,11 @@ func WithExponentialBackoff(initialBackoff time.Duration, webhookFn func() error
 	wait.ExponentialBackoff(backoff, func() (bool, error) {
 		err = webhookFn()
+		// these errors indicate a need to retry an authentication check
+		if apierrors.IsServerTimeout(err) || apierrors.IsTimeout(err) || apierrors.IsTooManyRequests(err) {
```
do these not trigger the lower check on a specified client delay?

smarterclayton replied:
Not necessarily. We don't actually honor the delay right now, but we have two classes of things:

1. errors that mean retry
2. an error or response (that doesn't automatically mean retry) that still asks you to retry for some other reason

deads2k commented:
lgtm

smarterclayton commented:
/approve no-issue
